src/System.Management.Automation/engine/parser/ast.cs

```diff
@@ -5342,6 +5342,215 @@ internal override AstVisitAction InternalVisit(AstVisitor visitor)
         #endregion Visitors
     }
 
+    /// <summary>
+    /// A command-oriented flow-controlled pipeline chain.
+    /// E.g. <c>npm build &amp;&amp; npm test</c> or <c>Get-Content -Raw ./file.txt || "default"</c>.
+    /// </summary>
+    public class StatementChainAst : StatementAst
+    {
+        /// <summary>
+        /// Create a new statement chain AST from two statements and an operator.
+        /// </summary>
+        /// <param name="lhsStatement">The statement to the left of the operator.</param>
+        /// <param name="rhsStatement">The statement to the right of the operator.</param>
+        /// <param name="chainOperator">The operator used.</param>
+        /// <param name="extent">The extent of the chained statement.</param>
+        public StatementChainAst(
+            IScriptExtent extent,
+            PipelineBaseAst lhsStatement,
+            StatementAst rhsStatement,
+            TokenKind chainOperator)
+            : base(extent)
+        {
+            if (lhsStatement == null)
+            {
+                throw new ArgumentNullException(nameof(lhsStatement));
+            }
+
+            if (rhsStatement == null)
+            {
+                throw new ArgumentNullException(nameof(rhsStatement));
+            }
+
+            if (chainOperator != TokenKind.AndAnd && chainOperator != TokenKind.OrOr)
+            {
+                throw new ArgumentException(nameof(chainOperator));
+            }
+
+            LhsStatement = lhsStatement;
+            RhsStatement = rhsStatement;
+            Operator = chainOperator;
+
+            SetParent(LhsStatement);
+            SetParent(RhsStatement);
+        }
+
+        /// <summary>
+        /// The left hand statement in the chain.
+        /// </summary>
+        public PipelineBaseAst LhsStatement { get; }
+
+        /// <summary>
+        /// The right hand statement in the chain.
+        /// </summary>
+        public StatementAst RhsStatement { get; }
+
+        /// <summary>
+        /// The chaining operator used.
+        /// </summary>
+        public TokenKind Operator { get; }
+
+        /// <summary>
+        /// Create a copy of this Ast.
+        /// </summary>
+        public override Ast Copy()
+        {
+            return new StatementChainAst(Extent, CopyElement(LhsStatement), CopyElement(RhsStatement), Operator);
+        }
+
+        internal override object Accept(ICustomAstVisitor visitor)
+        {
+            if (!(visitor is ICustomAstVisitor2 visitor2))
+            {
+                return null;
+            }
```

Comment:
Maybe reverse it? As
```
if (visitor is ICustomAstVisitor2 visitor2)
{
    return visitor2.VisitStatementChain(this);
}

return null;
```

Reply:
This is probably short enough to be a ternary expression if you wanted to>:
```csharp
return visitor is ICustomAstVisitor2
    ? visitor2.VisitStatementChain(this)
    : null;
```

Reply:
Another option:

```csharp
return (visitor as ICustomAstVisitor2)?.VisitStatementChain(this);
```

Reply:
The reason I picked the option is did is to follow the golden path down to the end of the method. As the reader proceeds, like the compiler, there is a series of type refinements where we deal with input we're not really interested in, and if you go to the bottom of the method you see the most common and intended usage of the method (like the conclusion of a paragraph) rather than an unhelpful `null`.

The coalesced version also looks good (even though I think `var y = x as T; if (y != null) { }` is used far too often in the PowerShell codebase)

Reply:
I personally think `!(visitor is ICustomAstVisitor2 visitor2)` is less readable than `visitor is ICustomAstVisitor2 visitor2`. It's more burden to scan the code.

Reply:
> The coalesced version also looks good (even though I think var y = x as T; if (y != null) { } is used far too often in the PowerShell codebase)

Agreed.  If you're storing the cast (or ultimately returning a `bool`) then an `is` assignment is preferable imo.